python/src/scipp/compat/mantid.py

```diff
@@ -197,21 +197,127 @@ def validate_and_get_unit(unit, allow_empty=False):
         return known_units[unit]
 
 
-def init_pos(ws):
-    nHist = ws.getNumberHistograms()
-    pos = np.zeros([nHist, 3])
-
+def _to_spherical(input):
+    input["r"] = sc.sqrt(input["x"].data * input["x"].data +
+                         input["y"].data * input["y"].data +
+                         input["z"].data * input["z"].data)
+    input["t"] = sc.acos(input["z"].data / input["r"].data)
+    input["p"] = input["t"].data.copy()
+    sc.atan2(input["y"].data, input["x"].data, input["p"].data)
+
+
+def rotation_matrix_from_vectors(vec1, vec2):
+    """ Find the rotation matrix that aligns vec1 to vec2
+    :param vec1: A 3d "source" vector
+    :param vec2: A 3d "destination" vector
+    :return mat: Tranform (3x3) which aligns it with vec2.
+    """
+    a, b = (vec1 /
+            np.linalg.norm(vec1)).reshape(3), (vec2 /
+                                               np.linalg.norm(vec2)).reshape(3)
+    v = np.cross(a, b)
+    c = np.dot(a, b)
+    s = np.linalg.norm(v)
+    if s == 0:
+        return np.eye(3)
+    kmat = np.array([[0, -v[2], v[1]], [v[2], 0, -v[0]], [-v[1], v[0], 0]])
+    rotation_matrix = np.eye(3) + kmat + kmat.dot(kmat) * ((1 - c) / (s**2))
+    return rotation_matrix
+
+
+def matrix_mult(pos, m):
+    if not pos.shape[0]:
+        return pos
+    m1 = sc.Variable(['y'],
+                     values=[np.array(m[0, :])],
+                     dtype=sc.dtype.vector_3_float64)
+    m2 = sc.Variable(['y'],
+                     values=[np.array(m[1, :])],
+                     dtype=sc.dtype.vector_3_float64)
+    m3 = sc.Variable(['y'],
+                     values=[np.array(m[2, :])],
+                     dtype=sc.dtype.vector_3_float64)
+    xn = sc.sum(sc.dot(m1, pos), 'y')
+    yn = sc.sum(sc.dot(m2, pos), 'y')
+    zn = sc.sum(sc.dot(m3, pos), 'y')
+    return sc.geometry.position(xn, yn, zn)
+
+
+def init_pos(ws, source_pos, sample_pos):
+    import numpy as np
     spec_info = ws.spectrumInfo()
-    for i in range(nHist):
-        if spec_info.hasDetectors(i):
-            p = spec_info.position(i)
-            pos[i, :] = [p.X(), p.Y(), p.Z()]
-        else:
-            pos[i, :] = [np.nan, np.nan, np.nan]
-    return sc.Variable(['spectrum'],
-                       values=pos,
-                       unit=sc.units.m,
-                       dtype=sc.dtype.vector_3_float64)
+
+    if sample_pos and source_pos:
+        det_info = ws.detectorInfo()
+        total_detectors = spec_info.detectorCount()
+        act_beam = (sample_pos - source_pos).values
+        rot = rotation_matrix_from_vectors(act_beam, [0, 0, 1])
+        inv_rot = rot.transpose()
+
+        pos_d = sc.Dataset()
+        # Create empty to hold position info for all spectra detectors
+        pos_d["x"] = sc.Variable(["Detector"],
```

Comment:
We have standardized to lower-case dimension names.
```suggestion
        pos_d["x"] = sc.Variable(["detector"],
```